src/coords.c

```diff
@@ -0,0 +1,463 @@
+
+#include <Python.h>
+#include <math.h>
+
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL pygeos_ARRAY_API
+#include "numpy/arrayobject.h"
+
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_3kcompat.h"
+
+#include "geos.h"
+#include "pygeom.h"
+
+static char get_coordinates(GEOSContextHandle_t, GEOSGeometry *, PyArrayObject *, npy_intp *);
+static void *set_coordinates(GEOSContextHandle_t, GEOSGeometry *, PyArrayObject *, npy_intp *);
+
+static char get_coordinates_simple(GEOSContextHandle_t context, GEOSGeometry *geom,
+                                   PyArrayObject *out, npy_intp *cursor) {
+    unsigned int n, i;
+    double *x, *y;
+    const GEOSCoordSequence *seq = GEOSGeom_getCoordSeq_r(context, geom);
+    if (seq == NULL) { return 0; }
+    if (GEOSCoordSeq_getSize_r(context, seq, &n) == 0) { return 0; }
+    for(i = 0; i < n; i++, *cursor += 1) {
+        x = PyArray_GETPTR2(out, *cursor, 0);
+        y = PyArray_GETPTR2(out, *cursor, 1);
+        if (GEOSCoordSeq_getX_r(context, seq, i, x) == 0) { return 0; }
+        if (GEOSCoordSeq_getY_r(context, seq, i, y) == 0) { return 0; }
+    }
+    return 1;
+}
+
+static char get_coordinates_polygon(GEOSContextHandle_t context, GEOSGeometry *geom,
+                                    PyArrayObject *out, npy_intp *cursor) {
+    int n, i;
+    GEOSGeometry *ring;
+
+    ring = (GEOSGeometry *) GEOSGetExteriorRing_r(context, geom);
+    if (ring == NULL) { return 0; }
+    if (!get_coordinates_simple(context, ring, out, cursor)) { return 0; }
+
+    n = GEOSGetNumInteriorRings_r(context, geom);
+    if (n == -1) { return 0; }
+    for(i = 0; i < n; i++) {
+        ring = (GEOSGeometry *) GEOSGetInteriorRingN_r(context, geom, i);
+        if (ring == NULL) { return 0; }
+        if (!get_coordinates_simple(context, ring, out, cursor)) { return 0; }
+    }
+    return 1;
+}
+
+static char get_coordinates_collection(GEOSContextHandle_t context, GEOSGeometry *geom,
+                                       PyArrayObject *out, npy_intp *cursor) {
+    int n, i;
+    GEOSGeometry *sub_geom;
+
+    n = GEOSGetNumGeometries_r(context, geom);
+    if (n == -1) { return 0; }
+    for(i = 0; i < n; i++) {
+        sub_geom = (GEOSGeometry *) GEOSGetGeometryN_r(context, geom, i);
+        if (sub_geom == NULL) { return 0; }
+        if (!get_coordinates(context, sub_geom, out, cursor)) { return 0; }
+    }
+    return 1;
+}
+
+static char get_coordinates(GEOSContextHandle_t context, GEOSGeometry *geom,
+                            PyArrayObject *out, npy_intp *cursor) {
+    int type = GEOSGeomTypeId_r(context, geom);
+    if ((type == 0) | (type == 1) | (type == 2)) {
+        return get_coordinates_simple(context, geom, out, cursor);
+    } else if (type == 3) {
+        return get_coordinates_polygon(context, geom, out, cursor);
+    } else if ((type >= 4) & (type <= 7)) {
+        return get_coordinates_collection(context, geom, out, cursor);
+    } else {
+        return 0;
```

Comment:
This could happen in case `GEOSGeomTypeId_r` errors (it could return -1).

I trust GEOS to do a callback to the errorhandler so that the error is set correctly.